acpi/src/lib.rs

```diff
@@ -115,170 +138,150 @@ where
+    /// Searches through the ACPI table headers and attempts to locate the table with a matching `T::SIGNATURE`.
+    pub fn find_table<T: AcpiTable>(&self) -> AcpiResult<PhysicalMapping<H, T>> {
+        use core::mem::size_of;
 
-        let header = sdt::peek_at_sdt_header(&result.handler, rsdt_address);
-        let mapping =
-            unsafe { result.handler.map_physical_region::<SdtHeader>(rsdt_address, header.length as usize) };
-
-        if revision == 0 {
-            /*
-             * ACPI Version 1.0. It's a RSDT!
-             */
-            mapping.validate(sdt::Signature::RSDT)?;
+        if self.revision == 0 {
+            let num_tables = ((self.mapping.length as usize) - size_of::<SdtHeader>()) / size_of::<u32>();
+            // ### Safety: Table pointer is known-good for these offsets and types.
+            let tables_base = unsafe { self.mapping.virtual_start().as_ptr().add(1).cast::<u32>() };
 
-            let num_tables = (mapping.length as usize - mem::size_of::<SdtHeader>()) / mem::size_of::<u32>();
-            let tables_base =
-                ((mapping.virtual_start().as_ptr() as usize) + mem::size_of::<SdtHeader>()) as *const u32;
+            for offset in 0..num_tables {
+                // ### Safety: Table pointer is known-good for these offsets and types.
+                let sdt_header_address = unsafe { tables_base.add(offset).read_unaligned() } as usize;
 
-            for i in 0..num_tables {
-                result.process_sdt(unsafe { *tables_base.add(i) as usize })?;
+                // ### Safety: `RSDT` guarantees its contained addresses to be valid.
+                let table_result = unsafe { read_table(self.handler.clone(), sdt_header_address) };
+                if table_result.is_ok() {
+                    return table_result;
+                }
             }
         } else {
-            /*
-             * ACPI Version 2.0+. It's a XSDT!
-             */
-            mapping.validate(sdt::Signature::XSDT)?;
-
-            let num_tables = (mapping.length as usize - mem::size_of::<SdtHeader>()) / mem::size_of::<u64>();
-            let tables_base =
-                ((mapping.virtual_start().as_ptr() as usize) + mem::size_of::<SdtHeader>()) as *const u64;
-
-            for i in 0..num_tables {
-                result.process_sdt(unsafe { *tables_base.add(i) as usize })?;
+            let num_tables = ((self.mapping.length as usize) - size_of::<SdtHeader>()) / size_of::<u64>();
+            // ### Safety: Table pointer is known-good for these offsets and types.
+            let tables_base = unsafe { self.mapping.virtual_start().as_ptr().add(1).cast::<u64>() };
+
+            for offset in 0..num_tables {
+                // ### Safety: Table pointer is known-good for these offsets and types.
+                let sdt_header_address = unsafe { tables_base.add(offset).read_unaligned() } as usize;
+
+                // ### Safety: `RSDT` guarantees its contained addresses to be valid.
```

Comment:
Very minor nit - it's an XSDT if we're dealing with `u64` addresses. I like keeping these signals around to help with reading the code, but if you can't be bothered to change it I'm not too concerned.